atlp-rwanda/ecommerce-app-predators-bn
# #9 Enable 2-Factor-Authentication

#### What does this PR do?
- Fixes #11
- Adds 2Factor Authentication feature using npm package **speakeasy**
- Adds configuration files directory at the root of the repository
- Adds **phone_number** property on the user model for sms verification if enabled
- Adds a swagger documentation of the 2Factor authentication API
#### Description of Task to be completed
- Document feature using swagger
- 2FA option is enabled on vendor login
- Appropriate error messages are returned on unsuccessful attempts
#### How should this be manually tested?
After successfully cloning and checkingout my branch:
- run `npm install`
- run `npm run start`
After starting the server open Postman and test the following routes:
1. `http://localhost:3000/auth/otp/generate`: Post request with a body in json containing property `user_id=<id>` replace **<id>** with the respective id you want to initiate 2fa on. On successful request you should get a response containing the following properties: `otp_base32, otp_url, qr_code_data_url`. Copy the `qr_code_data_url` value and paste it in the web browser and scan the given QR code with your authenticator app.

After connecting you authenticator, you should get OTP code which you can use on the following route.
**Alternative**: If you have a phone number set on the user you can go ahead and request a verifcation code via SMS by making a post request on this route `http://localhost:3000/auth/otp/getotp` and use that code on the following route.

2. `http://localhost:3000/auth/otp/verify`: Post request with a body containing properties `user_id, token` where **token** is the verification code from authenticator or SMS. It is cruicial to input code in time because each verification code expires after 30 seconds. On successful verification now your account will have 2Factor Authentication Enabled.

3. `http://localhost:3000/auth/otp/validate`: Post request with a body containing properties `user_id, token` where **token** is the verification code from authenticator or SMS. It is cruicial to input code in time because each verification code expires after 30 seconds. On successful validation, you will be granted access by setting the `otp_valid` property to true.

4. `http://localhost:3000/auth/otp/disable`: Post request with body containing property `user_id` of the user who is disabling 2fa. **NB:** That user should have 2fa enabled before disabling it.

#### Background Context
Before testing my PR you should have a registered user first.
For those who prefer verification codes via SMS, they should set their phone number first.

There are some environment variables needed to be set in the configuration files, including:

- DATABASE_URL
- otp_length
- otp_length_sms
- otp_name
- otp_issuer
- twilio:
      accountSid
      authToken
      phoneNumber

#### What are the relevant pivotal tracker stories?
- [Card #9](https://trello.com/c/VTYo8yUB) 
#### Screenshots
- Route 1: 
<img width="1298" alt="Screenshot 2023-04-14 at 20 34 25" src="https://user-images.githubusercontent.com/103200604/232129398-54921cc0-dc41-4a70-9ffd-394dc02b3949.png">

- QR code url in browser: 
<img width="1440" alt="Screenshot 2023-04-14 at 20 35 15" src="https://user-images.githubusercontent.com/103200604/232129697-d2a51023-7cb0-473b-89b4-2c2b63b0fd93.png">


- Route 2:
<img width="1306" alt="Screenshot 2023-04-14 at 20 36 33" src="https://user-images.githubusercontent.com/103200604/232129550-988090c2-b776-43f5-82bc-ab3949071745.png">

- Route 3: 
<img width="1296" alt="Screenshot 2023-04-14 at 20 37 02" src="https://user-images.githubusercontent.com/103200604/232129743-28e15a13-9d39-41a7-9c70-8fb495134618.png">

-Route 4:
<img width="1303" alt="Screenshot 2023-04-14 at 20 37 41" src="https://user-images.githubusercontent.com/103200604/232129790-6c5ef480-4752-47ff-8dbf-240eb7d71777.png">
